firestore/bulkwriter.go

```diff
@@ -0,0 +1,325 @@
+	write       *pb.Write             // the writes to apply to the database
+	attempts    int                   // number of times this write has been attempted
+	resultsLock sync.Mutex            // guards the cached wr and e values for the job
+	result      *WriteResult          // (cached) result from the operation
+	err         error                 // (cached) any errors that occurred
+	ctx         context.Context       // context for canceling/timing out results
+}
+
+// Results gets the results of the BulkWriter write attempt.
+// This method blocks if the results for this BulkWriterJob haven't been
+// received.
+func (j *BulkWriterJob) Results() (*WriteResult, error) {
+	j.resultsLock.Lock()
+	defer j.resultsLock.Unlock()
+	if j.result == nil && j.err == nil {
+		j.result, j.err = j.processResults() // cache the results for additional calls
+	}
+	return j.result, j.err
+}
+
+// processResults checks for errors returned from send() and packages up the
+// results as WriteResult objects
+func (j *BulkWriterJob) processResults() (*WriteResult, error) {
+	select {
+	case <-j.ctx.Done():
+		return nil, j.ctx.Err()
+	case bwr := <-j.resultChan:
+		if bwr.err != nil {
+			return nil, bwr.err
+		}
+		return writeResultFromProto(bwr.result)
+	}
+}
+
+// setError ensures that an error is returned on the error channel of BulkWriterJob.
+func (j *BulkWriterJob) setError(e error) {
+	bwr := bulkWriterResult{
+		err:    e,
+		result: nil,
+	}
+	j.resultChan <- bwr
+	close(j.resultChan)
+}
+
+// A BulkWriter supports concurrent writes to multiple documents. The BulkWriter
+// submits document writes in maximum batches of 20 writes per request. Each
+// request can contain many different document writes: create, delete, update,
+// and set are all supported.
+//
+// Only one operation (create, set, update, delete) per document is allowed.
+// BulkWriter cannot promise atomicity: individual writes can fail or succeed
+// independent of each other. Bulkwriter does not apply writes in any set order;
+// thus a document can't have set on it immediately after creation.
+type BulkWriter struct {
+	database        string           // the database as resource name: projects/[PROJECT]/databases/[DATABASE]
+	start           time.Time        // when this BulkWriter was started; used to calculate qps and rate increases
+	vc              *vkit.Client     // internal client
+	maxOpsPerSecond int              // number of requests that can be sent per second
+	docUpdatePaths  map[string]bool  // document paths with corresponding writes in the queue
+	limiter         rate.Limiter     // limit requests to server to <= 500 qps
+	bundler         *bundler.Bundler // handle bundling up writes to Firestore
+	ctx             context.Context  // context for canceling all BulkWriter operations
+	openLock        sync.Mutex       // guards against setting isOpen concurrently
+	isOpen          bool             // flag that the BulkWriter is closed
```

Comment:
nit: `open` to match `openLock` above.

Reply:
Ack. I prefer using 'verb+Noun' naming conventions for bool fields.